Gawron/ZPI-57-ZPI-5-BE-10-Get-cares-of-caretaker

Gawron97 opened the pull request:
Dodałem endpoint do wyszukiwania opiek opiekuna oraz możliwość filtrowania opiek.
Dodałem converter dla ZonedDateTime przy przyjmowania przez controller.
Dodałem annotacje, które: 

1.  Sprawdza, czy rola podana w headerze jest odpowiednia dla wywoływanego endpointu, jeśli nie otrzymujemy UnauthorizedException
2. Sprawdza, czy daty w klasie są odpowiednie, to znaczy data wcześniejsza rzeczywiście jest wcześniejsza, a data późniejsza jest późniejsza. Rozszerzyłem annotacje o możliwość podania wielu par dat do walidacji.

Gawron97 committed:
add general structure for getting caretaker cares.
add annotation to validate principal profile

Gawron97 committed:
extend filter by more fields

Gawron97 committed:
add converter from string to ZonedDateTime with proper formatt
throw custom exception in DateRangeValidator because of problems with default one

Gawron97 committed:
add CareService tests
add CareController tests
add validator and converter tests

Gawron97 committed:
change DateRangeValidator to validate more than one pair of dates

lukcyn commented on src/main/java/com/example/petbuddybackend/controller/CareController.java:
```diff
@@ -140,4 +150,29 @@ public CareDTO cancelCareByClient(@PathVariable Long careId,
+            @ApiResponse(responseCode = "400", description = "Invalid request data"),
+            @ApiResponse(responseCode = "403", description = "Authorized caretaker can only fetch cares"),
+            @ApiResponse(responseCode = "404", description = "When data provided is not found in the system")
+    })
```
```suggestion
    })
    @PreAuthorize("isAuthenticated()")
```

lukcyn commented on src/main/java/com/example/petbuddybackend/controller/CareController.java:
```diff
@@ -140,4 +150,29 @@ public CareDTO cancelCareByClient(@PathVariable Long careId,
+            @ApiResponse(responseCode = "404", description = "When data provided is not found in the system")
+    })
+    public Page<CareDTO> getCaretakerCares(
+            @RoleParameter @AcceptRole(acceptRole = "CARETAKER") @RequestHeader(value = "${header-name.role}") Role acceptRole,
```
```suggestion
            @RoleParameter @AcceptRole(acceptRole = Role.CARETAKER) @RequestHeader(value = "${header-name.role}") Role acceptRole,
```

Ja bym to zmodyfikował żeby to role przyjmowało po prostu

Gawron97 replied:
w sumie

lukcyn commented on src/main/java/com/example/petbuddybackend/utils/annotation/validation/AcceptRoleValidator.java:
```diff
@@ -0,0 +1,45 @@
+        }
+        return isValid;
+
+    }
```
```suggestion
    public boolean isValid(Role value, ConstraintValidatorContext context) {
        if(!isValid(value, acceptRoles)) {
            context.disableDefaultConstraintViolation();
            throw new UnauthorizedException(message);
        }
        return true;
    }
```

lukcyn commented on src/main/java/com/example/petbuddybackend/utils/annotation/validation/AcceptRoleValidator.java:
Generalnie zrobiłem jakiś szybki test i podmieniłem `String[] acceptRoles` na `Role[] acceptRoles` i normalnie spring przyjmuje enuma, więc imo lepiej nie validować stringów

Gawron97 replied:
tak tak przyjmuje, juz zmienilem

lukcyn commented on src/main/java/com/example/petbuddybackend/utils/converter/WebConfig.java:
```diff
@@ -0,0 +1,15 @@
+package com.example.petbuddybackend.utils.converter;
```
Do config/converter bym przeniósł

Gawron97 replied:
a rzeczywiscie

lukcyn commented on src/main/java/com/example/petbuddybackend/utils/converter/ZonedDateTimeConverter.java:
```diff
@@ -0,0 +1,19 @@
+package com.example.petbuddybackend.utils.converter;
```
Do config/converter bym przeniósł

lukcyn commented on src/main/java/com/example/petbuddybackend/dto/criteriaSearch/CareSearchCriteria.java:
```diff
@@ -0,0 +1,98 @@
+public record CareSearchCriteria(
+        @Schema(
+                description = "Filters by animal types in cares",
+                examples = {"DOG", "CAT", "BIRD", "SMALL_PET", "FISH", "REPTILE", "HORSE"},
```
Możliwe że te "examples" potrzebowały by update'a

Gawron97 replied:
wgl mam z tym problem bo to mi w swagerze zanych przykladow nie pokazuje

lukcyn commented on src/main/java/com/example/petbuddybackend/utils/specification/CareSpecificationUtils.java:
```diff
@@ -0,0 +1,187 @@
+        };
+    }
+
+    public static Specification<Care> addClientEmailFilter(String clientEmail) {
+        return (root, query, criteriaBuilder) -> {
+            Join<Care, Client> clientJoin = root.join(CLIENT);
+            return criteriaBuilder.equal(clientJoin.get(EMAIL), clientEmail);
+        };
+    }
```
Nieużywane nigdzie

Gawron97 replied:
jeszcze :)
z rozbiegu dodalem to do nastepnego taska: Zwracanie opiek klienta, czyli to samo ale z perspektywy klienta

lukcyn commented on src/main/java/com/example/petbuddybackend/utils/specification/CareSpecificationUtils.java:
```diff
@@ -0,0 +1,187 @@
+        };
+    }
+
+    public static Specification<Care> addCaretakerEmailsFilter(Set<String> caretakerEmails) {
+        return (root, query, criteriaBuilder) -> {
+            Join<Care, Caretaker> caretakerJoin = root.join(CARETAKER);
+            return caretakerJoin.get(EMAIL).in(caretakerEmails);
+        };
+    }
```
Nieużywane nigdzie

lukcyn commented on src/main/java/com/example/petbuddybackend/controller/CareController.java:
```diff
@@ -140,4 +150,29 @@ public CareDTO cancelCareByClient(@PathVariable Long careId,
+            @ApiResponse(responseCode = "404", description = "When data provided is not found in the system")
+    })
+    public Page<CareDTO> getCaretakerCares(
+            @RoleParameter @AcceptRole(acceptRole = "CARETAKER") @RequestHeader(value = "${header-name.role}") Role acceptRole,
```
W ogóle to w przypadku braku tego headera nie rzuca UnauthorizedException tylko MissingHeaderException

Gawron97 replied:
wg mnie to dobrze, bo to rzeczywiscie sie wydarzylo

lukcyn commented on src/test/java/com/example/petbuddybackend/controller/CareControllerIntegrationTest.java:
```diff
@@ -493,5 +498,63 @@ void cancelCareByClient_WhenCareIsAlreadyAcceptedByCaretaker_ShouldThrowIllegalA
+
+    @Test
+    @WithMockUser(username = "caretakerEmail")
+    void getCaretakerCares_whenRoleIsNotProvided_ShouldThrowUnauthorizedException() throws Exception {
+        // Given
+        PersistenceUtils.addCare(careRepository, caretaker, client, animalRepository.findById("DOG").get());
+        // When and Then
+        mockMvc.perform(get("/api/care/caretaker-cares")
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isBadRequest());
+    }
```
ShouldThrowUnauthorizedException a sprawdzane jest isBadRequest

Gawron97 replied:
a bo to z tym headerem, ze rzuca co innego nie Unauthorized

lukcyn commented on src/main/java/com/example/petbuddybackend/controller/CareController.java:
```diff
@@ -140,4 +150,29 @@ public CareDTO cancelCareByClient(@PathVariable Long careId,
+            @RoleParameter @AcceptRole(acceptRole = "CARETAKER") @RequestHeader(value = "${header-name.role}") Role acceptRole,
+            @TimeZoneParameter @RequestHeader(value = "${header-name.timezone}", required = false) String timeZone,
+            @ParameterObject @ModelAttribute @Valid SortedPagingParams pagingParams,
+            @ParameterObject @ModelAttribute @Valid CareSearchCriteria filters,
```
Być może przydałby się jakiś test do sortowania, żeby sprawdzić czy polami z CareDTO możemy sortować Care. Byłoby to o tyle przydatne, że jakbyśmy zmieniali nazwy pól to od razu byłoby to widoczne. Poza tym nie jestem pewny czy możemy sortować po `String animal` z CareDTO bo Care ma `Animal animal`

Dla opiekunów robiłem taki test:
```java
    @Test
    void testGetCaretakers_sortingParamsShouldAlignWithDTO() {
        List<String> fieldNames = ReflectionUtils.getPrimitiveNames(CaretakerDTO.class);
        fieldNames.addAll(getPrimitiveNames(AddressDTO.class, "address_"));
        fieldNames.addAll(getPrimitiveNames(AccountDataDTO.class, "accountData_"));

        for(String fieldName : fieldNames) {
            assertDoesNotThrow(() -> caretakerService.getCaretakers(
                    PageRequest.of(0, 10, Sort.by(Sort.Direction.ASC, fieldName)),
                    CaretakerSearchCriteria.builder().build()
            ));
        }
    }
``` 

I imo nie trzeba koniecznie sprawdzać czy strona jest na pewno posortowana, ale tylko czy operacja się nie wywali

Gawron97 committed:
move converters to config
change AcceptRole annotation to accept Enum instead of String
add security for getCaretakerCares
add test for sorting cares

Gawron97 committed:
add more swagger documentation for getCaretakerCares
